python-checks/src/main/java/org/sonar/python/checks/TorchModuleModeShouldBeSetAfterLoadingCheck.java

```diff
@@ -0,0 +1,114 @@
+  public static final int IS_TORCH_LOAD_CALL_MAX_RECURSIVE_COUNTER = 10;
+
+  private ReachingDefinitionsAnalysis reachingDefinitionsAnalysis;
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FILE_INPUT, ctx -> reachingDefinitionsAnalysis =
+      new ReachingDefinitionsAnalysis(ctx.pythonFile()));
+
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, ctx -> {
+      CallExpression callExpr = (CallExpression) ctx.syntaxNode();
+      List<Usage> receiverUsages = getForwardUsages(callExpr);
+      if (isLoadStateDictCall(callExpr) && !hasEvalOrTrainUsage(receiverUsages) && !isModelPassedOn(receiverUsages)) {
+        ctx.addIssue(callExpr.callee(), MESSAGE);
+      }
+    });
+  }
+
+  private boolean isLoadStateDictCall(CallExpression callExpr) {
+    // To properly check if the correct load_state_dict is called, typeshed type information would be required.
+    // Since this is currently not possible, we check if the parameter to load_state_dict is torch.load(...),
+    // with the assumption that if torch.load is passed to this load_state_dict, it is probably the correct method
+    if(callExpr.callee() instanceof QualifiedExpression qualifiedExpr) {
+      return LOAD_STATE_DICT_NAME.equals(qualifiedExpr.name().name()) && containsTorchLoadCall(callExpr.arguments());
+    }
+    return false;
+  }
+
+  private boolean containsTorchLoadCall(List<Argument> args) {
+    return args.stream()
+      .flatMap(TreeUtils.toStreamInstanceOfMapper(RegularArgument.class))
+      .anyMatch(arg -> isTorchLoadCall(arg.expression(), 0));
+  }
+
+  private boolean isTorchLoadCall(Expression expr, int recursiveCounter) {
+    if (recursiveCounter > IS_TORCH_LOAD_CALL_MAX_RECURSIVE_COUNTER) {
+      return false;
+    } else if (expr instanceof CallExpression callExpr) {
+      Symbol calleeSymbol = callExpr.calleeSymbol();
+      return calleeSymbol != null && TORCH_LOAD_FQN.equals(calleeSymbol.fullyQualifiedName());
+    } else if (expr instanceof Name name) {
+      return reachingDefinitionsAnalysis.valuesAtLocation(name).stream()
+        .anyMatch(definitionExpr -> isTorchLoadCall(definitionExpr, recursiveCounter + 1));
+    } else {
+      return false;
+    }
+  }
+
+  private static List<Usage> getForwardUsages(CallExpression callExpr) {
+    List<Usage> usages = getFunctionCallReceiverName(callExpr)
+      .map(name -> name.symbol().usages())
+      .orElse(Collections.emptyList());
+
+    return usages.stream()
+      .filter(usage -> usage.tree().firstToken().line() > callExpr.firstToken().line())
+      .toList();
+  }
+
+  private static Optional<Name> getFunctionCallReceiverName(CallExpression callExpr) {
+    Expression calleeExpr = callExpr.callee();
+    if (calleeExpr instanceof QualifiedExpression qualifiedExpr) {
+      return Optional.of(qualifiedExpr.qualifier()).flatMap(TreeUtils.toOptionalInstanceOfMapper(Name.class));
+    }
+    return Optional.empty();
```

Comment:
Here maybe it would be preferable to directly start the check with an optional. I find it makes it easier to read.
```
Optional.of(callExpr.callee())
  .flatMap(toOptionalInstanceOfMapper(QualifiedExpression)
...
```